tensorflow/lite/micro/tools/benchmarking/metrics.cc

```diff
@@ -15,88 +15,299 @@ limitations under the License.
 
 #include "tensorflow/lite/micro/tools/benchmarking/metrics.h"
 
-#include <sys/types.h>
-
 #include <cstddef>
+#include <cstring>
+#include <iterator>
+#include <type_traits>
+#include <utility>
 
-#include "tensorflow/lite/micro/tools/benchmarking/log_utils.h"
+#include "tensorflow/lite/micro/micro_log.h"
 
 namespace tflite {
 
-void LogArenaAllocations(
-    const tflite::RecordingSingleArenaBufferAllocator* allocator,
-    const PrettyPrintType type) {
-  constexpr int kArenaRows = 3;
-  constexpr int kArenaCols = 3;
+namespace {
 
-  const size_t total_bytes = allocator->GetUsedBytes();
+struct LogArenaRecord {
+  const char* title;
+  int allocations;
+  float percentage;
+};
 
-  size_t allocations[kArenaRows] = {total_bytes,
-                                    allocator->GetNonPersistentUsedBytes(),
-                                    allocator->GetPersistentUsedBytes()};
-  char titles[kArenaRows][kMaxStringLength] = {"Total", "Head", "Tail"};
-  char headers[kArenaRows][kMaxStringLength] = {"Arena", "Bytes", "% Arena"};
+struct LogAllocationRecord {
+  const char* title;
+  int type;
+  int used_bytes;
+  int requested_bytes;
+  int count;
+  float percentage;
+};
 
-  char data[kArenaCols][kArenaRows][kMaxStringLength];
-  for (int i = 0; i < kArenaRows; ++i) {
-    MicroStrcpy(data[0][i], titles[i]);
-    FormatNumber<int32_t>(data[1][i], allocations[i]);
-    FormatAsPercentage(data[2][i], static_cast<int64_t>(allocations[i]),
-                       static_cast<int64_t>(total_bytes), 2);
-  }
+constexpr int kArenaRows = 3;
+constexpr int kArenaColumns = 3;
+
+constexpr int kAllocationTypes = 7;
+constexpr int kAllocationColumns = 6;
 
-  PrintFormattedData<kArenaRows, kArenaCols>(headers, data, kArenaRows, type,
-                                             "Arena");
+constexpr int kMaxBufSize = 100;
+
+LogArenaRecord GetLogArenaRecord(
+    const tflite::RecordingSingleArenaBufferAllocator* allocator,
+    int row_index) {
```

Comment:
We should probably add a DCHECK that row_index < kArenaRows

Reply:
done